src/physics/initial_conditions/initial_condition_function.cpp

```diff
@@ -352,6 +352,103 @@ ::value(const dealii::Point<dim,real> &point, const unsigned int /*istate*/) con
+    const double beta = M_infty * 5 * sqrt(2.0)/4.0/pi * exp(1.0/2.0);
+    const double alpha = pi/4; //rad
+
+    // Centre of the vortex  at t=0
+    const double x0 = 0.0;
+    const double y0 = 0.0;
+    const double x = point[0]-x0;
+    const double y = point[1] - y0;
+
+    const double Omega = beta * exp(-0.5/sigma/sigma* (x/R * x/R + y/R * y/R));
+    const double delta_Ux = -y/R * Omega;
+    const double delta_Uy =  x/R * Omega;
+    const double delta_T  = -(gam-1.0)/2.0 * Omega * Omega;
+
+    // Primitive
+    const double rho = pow((1 + delta_T), 1.0/(gam-1.0));
+    const double Ux = M_infty * cos(alpha) + delta_Ux;
+    const double Uy = M_infty * sin(alpha) + delta_Uy;
+    const double Uz = 0;
+    const double p = 1.0/gam*pow(1+delta_T, gam/(gam-1.0));
+
+    //Convert to conservative variables
+    if (istate == 0)      return rho;       //density
+    else if (istate == nstate-1) return p/(gam-1.0) + 0.5 * rho * (Ux*Ux + Uy*Uy + Uz*Uz);   //total energy
+    else if (istate == 1) return rho * Ux;  //x-momentum
+    else if (istate == 2) return rho * Uy;  //y-momentum
+    else if (istate == 3) return rho * Uz;  //z-momentum
+    else return 0;
+
+}
+
+// ========================================================
+// KELVIN-HELMHOLTZ INSTABILITY
+// See Chan et al., On the entropy projection..., 2022, Pg. 15
+//     Note that some equations are not typed correctly
+//     See github.com/trixi-framework/paper-2022-robustness-entropy-projection
+//     for initial condition which is implemented herein
+// ========================================================
+template <int dim, int nstate, typename real>
+InitialConditionFunction_KHI<dim,nstate,real>
+::InitialConditionFunction_KHI (const double atwood_number_input)
+        : InitialConditionFunction<dim,nstate,real>()
+        , atwood_number(atwood_number_input)
+{
+    // Nothing to do here yet
+}
+
+template <int dim, int nstate, typename real>
+inline real InitialConditionFunction_KHI<dim,nstate,real>
+::value(const dealii::Point<dim,real> &point, const unsigned int istate) const
+{
+    const double pi = dealii::numbers::PI;
+    const double gam = 1.4;
+    
+    const double B = 0.5 * (tanh(15*point[1] + 7.5) - tanh(15*point[1] - 7.5));
+
+    const double rho1 = 0.5;
+    const double rho2 = rho1 * (1 + atwood_number) / (1 - atwood_number);
+
+    const double density = rho1 + B * (rho2-rho1);
+    const double pressure = 1;
+    const double x_velocity = B - 0.5;
+    const double y_velocity = 0.1 * sin(2 * pi * point[0]);
+    const double vel2 = x_velocity*x_velocity + y_velocity*y_velocity;
+
+    //Convert to conservative variables
+    if (istate == 0)      return density;       //density 
+    else if (istate == nstate-1) return pressure/(gam-1.0) 
+        + 0.5 * density * vel2;   //total energy
+    else if (istate == 1) return density * x_velocity;  //x-momentum
+    else if (istate == 2) return density * y_velocity;  //y-momentum
```

Comment:
could unify this with `convert_primitive_to_conversative_value` from `InitialConditionFunction_TaylorGreenVortex` by moving the lines in the photo attached to some function in the base class `InitialConditionFunction` and pass all  the primitive variables as the arguments 
<img width="845" alt="image" src="https://user-images.githubusercontent.com/43619715/222925079-dee822f0-da87-42e7-8af7-91500dfc7c0f.png">

Reply:
I made this change in the latest push. I checked that the KHI test and the split forms TGV tests behave the same as they did before.